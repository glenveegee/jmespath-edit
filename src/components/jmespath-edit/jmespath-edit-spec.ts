import { newSpecPage, SpecPage } from '@stencil/core/testing';
import { JmespathEdit } from './jmespath-edit';

describe('jmespath-edit', () => {
  // let rootInst: JmespathEdit;
  let rootEl: HTMLJmespathEditElement;
  let page: SpecPage;

  beforeEach(async () => {
    page = await newSpecPage({
      components: [JmespathEdit],
      html: '<jmespath-edit></jmespath-edit>',
    });
    // rootInst = page.rootInstance;
    rootEl = page.root as HTMLJmespathEditElement;
  });

  describe('sanity', () => {
    it('builds', () => {
      expect(new JmespathEdit()).toBeTruthy();
    });
  });

  describe('render', () => {
    it('renders default datasource and query', () => {
      expect((rootEl.querySelector('.expression div input') as HTMLInputElement).value).toEqualText(`locations[?state == 'WA'].name | sort(@) | {WashingtonCities: join(', ', @)}`)
//       expect((rootEl.querySelector('.results .input div textarea') as HTMLTextAreaElement)).toEqualLightHtml(`<textarea value=\"{
//   &quot;locations&quot;: [
//     {
//       &quot;name&quot;: &quot;Seattle&quot;,
//       &quot;state&quot;: &quot;WA&quot;
//     },
//     {
//       &quot;name&quot;: &quot;New York&quot;,
//       &quot;state&quot;: &quot;NY&quot;
//     },
//     {
//       &quot;name&quot;: &quot;Bellevue&quot;,
//       &quot;state&quot;: &quot;WA&quot;
//     },
//     {
//       &quot;name&quot;: &quot;Olympia&quot;,
//       &quot;state&quot;: &quot;WA&quot;
//     }
//   ]
// }\"></textarea>`)
//       expect((rootEl.querySelector('.results .output div pre') as HTMLPreElement)).toEqualText(`{
//   "WashingtonCities": "Bellevue, Olympia, Seattle"
// }`)
    });
  });
});
