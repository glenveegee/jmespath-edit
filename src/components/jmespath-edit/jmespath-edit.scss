:host {
  /* Component styles go here */
  padding: 1%;
  width: 98%;
  display: block;

  section {

    flex: 1;
    & h2 {
      font-family: sans-serif;
    }

    & div * {
      font-family: Consolas, Monaco, Menlo, 'Courier New', monospace;
      font-size: 14px;
    }
  }

  & .expression div {
    margin-bottom: 20px;
    width: inherit;

    & input, textarea {
      width: 100%;
      padding: 1%;
      box-sizing: border-box;
    }
  }

  & .results {
    display: flex;
    flex: 1 1 auto;
    flex-wrap: wrap;
    flex-direction: row;

    & section {

      & div {
        // max-width: min-content;
        // min-width: max-content;
        // padding: 0 0.5rem;
        width: calc(100% - 0.5rem);
        height: 60vh;
        box-sizing: border-box;

        & pre,
        & textarea {
          -webkit-box-sizing: border-box;
          -moz-box-sizing: border-box;
          box-sizing: border-box;
          overflow: auto;
          width: inherit;
          height: inherit;
          padding: 0.5rem;
        }

        .inputWarning {
          padding: 0rem 0 0.3rem 1rem;
          border-left: 4px solid goldenrod;
          color: goldenrod;
          max-height: 1rem;
        }
      }
    }
  }

  & .output div {
    // padding: 1vw;
    background-color: #eee;
    // box-sizing: content-box;
  }
}
