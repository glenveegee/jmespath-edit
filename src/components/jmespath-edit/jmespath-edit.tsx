import { Component, ComponentInterface, State, Host, h, Prop, Watch } from '@stencil/core';

import {BehaviorSubject, combineLatest, Subscription} from 'rxjs';
import {query} from '../../utils/metrichor/jmespath-plus';
import { JSONValue } from '@metrichor/jmespath/dist/types/typings';

@Component({
  styleUrl: 'jmespath-edit.scss',
  tag: 'jmespath-edit',
  shadow: true,
})
export class JmespathEdit implements ComponentInterface {

  expression$ = new BehaviorSubject<string>('');
  source$ = new BehaviorSubject<any>(null);
  query$ = combineLatest([this.expression$, this.source$]);

  listener!: Subscription;
  @Prop() expression = '';
  @Prop() json: JSONValue = null;

  @State() output = '';
  @State() inputError = '';

  @Watch('expression')
  updateExpressionHandler(newExpression: string) {
    if (newExpression) {
      this.expression$.next(newExpression)
    }
  }

  @Watch('json')
  updateSourceHandler(newSource: JSONValue) {
    if (newSource) {
      this.source$.next(newSource)
    }
  }

  runLibrarySpecificQuery = async (expression: string, source: JSONValue): Promise<JSONValue> => {
    return query(expression, source)
  }

  runQuery = async ([expression, source]: [string, JSONValue]) => {
    if (!expression || !source) return
    try {
      const result = await this.runLibrarySpecificQuery(expression, source);
      this.output = JSON.stringify(result, null, 2)
    } catch (error) {
      this.output = error.message
    }
  }

  coerceJSON = (json: JSONValue): JSONValue => {
    if (!json) return json;
    if (typeof json === 'string') {
      try {
        return JSON.parse(json)
      } catch (error) {
        return json
      }
    }
    return json;
  }

  componentWillLoad() {
    this.listener = this.query$.subscribe(this.runQuery)
    this.expression$.next(this.expression);
    this.source$.next(this.coerceJSON(this.json));
  }

  componentDidUnload() {
    this.listener.unsubscribe()
  }

  setSource = (e: any) => {
    const source = e.target.value;
    if (!source) return;
    try {
      this.source$.next(JSON.parse(source))
      this.inputError = '';
    } catch (error) {
      console.error(error)
      this.inputError = error.message
    }
  }

  setExpression = (e: any) => {
    const expression = e.target.value;
    if (expression) {
      this.expression$.next(expression)
    }
  }

  render() {

    const currentExpression = this.expression$.getValue();
    const currentSource = this.source$.getValue();

    return (
      <Host>
        <section class="expression" part="jmespath-expression">
          <h2>EXPRESSION</h2>
          <div>
            <input type="text" value={currentExpression}
            onInput={this.setExpression}
            />
          </div>
        </section>
        <div class="results" part="jmespath-expression-results">
          <section class="input" part="jmespath-expression-results-input">
            <h2>SOURCE DATA</h2>
            <div>
              {
                this.inputError && <div class='inputWarning'>{this.inputError}</div> || null
              }
              <textarea value={currentSource && JSON.stringify(currentSource, null, 2) || ''} onInput={this.setSource} />
            </div>
          </section>
          <section class="output" part="jmespath-expression-results-output">
            <h2>OUTPUT</h2>
            <div>
              <pre>{this.output}</pre>
            </div>
          </section>
        </div>
      </Host>
    );
  }
}
