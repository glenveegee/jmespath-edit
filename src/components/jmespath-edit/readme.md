# jmespath-edit



<!-- Auto Generated Below -->


## Properties

| Property     | Attribute    | Description | Type                                                                             | Default                      |
| ------------ | ------------ | ----------- | -------------------------------------------------------------------------------- | ---------------------------- |
| `expression` | `expression` |             | `string`                                                                         | `''`                         |
| `json`       | `json`       |             | `JSONValue[] \| boolean \| number \| string \| { [member: string]: JSONValue; }` | `null`                       |
| `library`    | `library`    |             | `string`                                                                         | `'@metrichor/jmespath-plus'` |


## Dependencies

### Used by

 - [jmespath-edit-demo](../jmespath-edit-demo)

### Graph
```mermaid
graph TD;
  jmespath-edit-demo --> jmespath-edit
  style jmespath-edit fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
