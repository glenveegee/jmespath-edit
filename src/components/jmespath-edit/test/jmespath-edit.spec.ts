import { newSpecPage, SpecPage } from '@stencil/core/testing';
import { JmespathEdit } from '../jmespath-edit';

describe('jmespath-edit', () => {
  let rootEl: HTMLJmespathEditElement;
  let page: SpecPage;

  beforeEach(async () => {
    page = await newSpecPage({
      components: [JmespathEdit],
      html: '<jmespath-edit></jmespath-edit>',
    });
    rootEl = page.root as HTMLJmespathEditElement;
  });

  describe('render', () => {
    it('renders default datasource and query', () => {
      expect(rootEl).toMatchSnapshot();
    })

    it('renders default datasource and query', async () => {
      rootEl.expression = "locations[?state == 'WA'].name | sort(@) | {WashingtonCities: join(', ', @)}";
      rootEl.json = {"locations": [{"name": "Seattle", "state": "WA"},{"name": "New York", "state": "NY"},{"name": "Bellevue", "state": "WA"},{"name": "Olympia", "state": "WA"}]};
      await page.waitForChanges();

      expect((rootEl.shadowRoot.querySelector('.expression div input') as HTMLInputElement)).toMatchSnapshot()
      expect((rootEl.shadowRoot.querySelector('.results .output div pre') as HTMLPreElement)).toEqualText(`
{
  \"WashingtonCities\": \"Bellevue, Olympia, Seattle\"
}`)
    });
  });
});
