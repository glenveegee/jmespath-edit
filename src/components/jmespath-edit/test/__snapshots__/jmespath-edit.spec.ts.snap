// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jmespath-edit render renders default datasource and query 1`] = `
<jmespath-edit>
  <mock:shadow-root>
    <section class="expression" part="expression">
      <h2>
        EXPRESSION
      </h2>
      <div><textarea value=""></textarea>
      </div>
    </section>
    <div class="results" part="results">
      <section class="input" part="results-input">
        <h2>
          SOURCE DATA
        </h2>
        <div><textarea value=""></textarea>
        </div>
      </section>
      <section class="output" part="results-output">
        <h2>
          OUTPUT
        </h2>
        <div><pre></pre>
        </div>
      </section>
    </div>
  </mock:shadow-root>
</jmespath-edit>
`;

exports[`jmespath-edit render renders default datasource and query 2`] = `null`;
